### PR TITLE
first attempt at supporting `opm build-install`

### PR DESCRIPTION
**Pull request by jvanasco**
This is my first attempt at handling https://github.com/openresty/opm/issues/39  

I haven't touched Perl in years so apologies in advance.

It seems to work, but there aren't any tests to plug into.  I documented new functions inline, and also added docs for existing functions too.

Notes: 

* The command is invoked as `opm build-install` as it wraps the build process and installs locally (as requested)

* A new core subroutine of `do_build_install` was created, which handles most of the overall logic

* build-install will always remove an existing version, it will not compare revisions. 

* Variable names were standardized for easier maintenance:

   * `$rcfile` -> `$rc_file` (same style as `$ini_file`)
   * `$data` -> `$rc_data`, `$ini_data`
   * `$default_sec` -> `$rc_default_sec`, `$ini_default_sec`

* I noted what the arguments meant to various subroutines, both when invoked and in the subroutine iteself

I tried to leverage existing code, and am not thrilled with my approach (but it does work)

* `install_target` was dropped into two routines: `install_target` and `install_distribution`.  
* `install_distribution` handles installing the actual distribution file.  Most metadata is calculated in `install_target` or `do_build_install`
* in order to keep track of some package and directory data, I created a secondary context object with `init_package_ctx`.  It's used alongside the existing `init_installation_ctx`, but is limited to the current package info, and mostly used for specifying the cache directory vs the current directory.

### CODE REVIEW

File: bin/opm [hunk 1]
```diff
@@ -661,6 +674,67 @@ _EOC_
     #if ($server_build) {
     #err "something bad bad bad.\n";
     #}
+
+    warn "build successful: '$TarBall'\n";
+}
+
+sub do_build_install () {
+=pod
```
Comment (agentzh) [inline bin/opm]: These comments should be comments instead of POD documentation. Also, they should appear before the sub prototype.

File: bin/opm [hunk 2]
```diff
@@ -661,6 +674,67 @@ _EOC_
...
+	# phase 1 - `opm build`
+	do_build(0);  # 0 = normal
+
+	# phase 2 - `cd $dist_name-$version/`
+	# phase 2a - derive the version data	
+    my $dist_file = "dist.ini";
+    my $ini_data = read_ini($dist_file);
+    my $ini_default_sec = $ini_data->{default};
+    my $dist_name = $ini_default_sec->{name};
+    my $version = $ini_default_sec->{version};
```
Comment (agentzh) [inline bin/opm]: Bad indentation?

File: bin/opm [hunk 3]
```diff
@@ -183,30 +197,29 @@ sub do_build ($) {
 
     my $account;
     if (!$server_build) {
-        my ($rcfile, $data) = get_rc_file();
-        my $default_sec = $data->{default};
-        $account = delete $default_sec->{github_account};
+        my ($rc_file, $rc_data) = get_rc_file();
+        my $rc_default_sec = $rc_data->{default};
```
Comment (agentzh) [inline bin/opm]: Please avoid such unrelated changes. I prefer minimal change sets in each PR. Thank you.

File: bin/opm [hunk 4]
```diff
@@ -165,15 +172,22 @@ sub do_get {
...
 
 sub do_build ($) {
-    my $server_build = shift;
+=pod
+`do_build($)`
+The first argument indicates the build type
+	0 = build
+	1 = server-build
+=cut
+    my $_build_type = shift;
```
Comment (agentzh) [inline bin/opm]: Please avoid local variable names led by `_`.

File: bin/opm [hunk 5]
```diff
@@ -165,15 +172,22 @@ sub do_get {
...
+    my $server_build = ($_build_type == 1) ? 1 : 0;
```
Comment (agentzh) [inline bin/opm]: Please avoid magical numbers in the build type values.

File: bin/opm [hunk 6]
```diff
@@ -250,14 +263,14 @@ sub do_build ($) {
...
-    my $dist_abstract = delete $default_sec->{abstract};
+    my $dist_abstract = delete $ini_default_sec->{abstract};
     if (!$dist_abstract) {
         err "$dist_file: key \"abstract\" not found in the default section.\n";
     }
 
     check_utf8_field($dist_file, 'abstract', $dist_abstract);
 
-    my $repo_link = delete $default_sec->{repo_link};
+    my $repo_link = delete $ini_default_sec->{repo_link};
```
Comment (agentzh) [inline bin/opm]: Please avoid such unrelated changes. Thank you.

File: bin/opm [hunk 7]
```diff
@@ -1345,23 +1419,56 @@ sub install_target ($$) {
...
+    return;
+
+SKIP_INSTALL:
+
+    warn "Package $name-$installed_ver already installed.\n";
+    return;
+}
+
+
+sub install_distribution($$$){
```
Comment (agentzh) [inline bin/opm]: Need a space before `{` and `(`.

File: bin/opm [hunk 8]
```diff
@@ -1909,20 +2010,53 @@ sub do_upgrade {
...
+Core installation context
+Single argument is boolean `$upgrade`
+
+	Called by		   Upgrade?
+	* do_get           0
+	* do_build         0 
+	* do_build_install 0
+	* do_remove        0 
+	* do_upgrade       1
+	* do_update        1
```
Comment (agentzh) [inline bin/opm]: Remaining this back-reference list is quite a burden. Better avoid it.

File: bin/opm (hunk 2 above)
Comment (agentzh) [inline bin/opm]: We should use 4 space indentation consistently.

File: bin/opm (hunk 3 above)
Comment (jvanasco) [inline bin/opm]: Would you mind if I do a PR first that standardizes the variable names?  This was a bit of a headache to unwind, because `$data` and `$default_sec` would be recycled from the rc file to the ini file.

**Comment by jvanasco** [comment]
I'll work on all the changes tomorrow.  Sorry for the whitespace issues, my editor didn't pick up Perl and auto-expand the tabs (it does for .py and .pl files)

**Comment by jvanasco** [comment]
I had to do a manual rebase and force-push, but I've got this working again.

File: bin/opm [hunk 9]
```diff
@@ -658,9 +671,81 @@ _EOC_
...
+This builds the currently authored package, then installs it locally (without
+communicated with the opm server).
+
+It performs these functions:
+    opm build
+    cd $dist_name-$version/
+    opm server-build
+    install $dist_name-$version.opm.tar.gz  # psuedocode
+=end comment
+=cut
```
Comment (agentzh) [inline bin/opm]: Better put blank lines around POD directives.

File: bin/opm [hunk 10]
```diff
@@ -658,9 +671,81 @@ _EOC_
...
+    my $meta_file = File::Spec->catfile($ctx->{manifest_dir}, "$dist_name.meta");
+    my $remove_old;
+    if (-f $meta_file) {
+    	# for `build-install`:
+    	#     *always* `remove_old` if the `meta_file` exists, regardless of version
+    	#     however, `get` will still need to compare version numbers
+    	$remove_old = 1;
+    }
+
+	# package_ctx is needed to get the directories right for `install_distribution`
```
Comment (agentzh) [inline bin/opm]: Bad indentation?

File: bin/opm [hunk 11]
```diff
@@ -12,7 +12,7 @@ use File::Find ();
 use File::Path qw( make_path );
 use File::Spec ();
 use Config ();
-use Cwd qw( realpath cwd );
+use Cwd qw( realpath cwd getcwd );
```
Comment (agentzh) [inline bin/opm]: Why import `getcwd`? Why not just use `cwd`?

File: bin/opm [hunk 12]
```diff
@@ -658,9 +671,81 @@ _EOC_
...
+    # phase 2a - derive the version data
+    my $dist_file = "dist.ini";
+    my $data = read_ini($dist_file);
+    my $default_sec = $data->{default};
+    my $dist_name = $default_sec->{name};
+    my $version = $default_sec->{version};
+
+    # phase 2b - `cd `
+    my $versioned_directory = "$dist_name-$version";
+    chdir $versioned_directory;
```
Comment (agentzh) [inline bin/opm]: Should check if `chdir` fails.

File: bin/opm [hunk 13]
```diff
@@ -658,9 +671,81 @@ _EOC_
...
+    my $ctx = init_installation_ctx(0);  # 0 = do not `$upgrade`
+
+    # is it installed?
+    my $meta_file = File::Spec->catfile($ctx->{manifest_dir}, "$dist_name.meta");
+    my $remove_old;
+    if (-f $meta_file) {
+    	# for `build-install`:
+    	#     *always* `remove_old` if the `meta_file` exists, regardless of version
+    	#     however, `get` will still need to compare version numbers
+    	$remove_old = 1;
```
Comment (agentzh) [inline bin/opm]: We should use 4-space indentations.

File: bin/opm [hunk 14]
```diff
@@ -1345,14 +1430,50 @@ sub install_target ($$) {
...
+    return;
+}
+
+
+=begin comment
+`install_distribution`
+This installs a distribution file (.tar.gz)
+=end comment
+=cut
+sub install_distribution($$$) {
```
Comment (agentzh) [inline bin/opm]: Style: need a space before `(`.

File: bin/opm [hunk 15]
```diff
@@ -1406,7 +1527,7 @@ sub install_target ($$) {
         err "no library files found in $dist_dir/.\n";
     }
 
-    if ($remove_old) {
+    if ($package_ctx->{remove_old}) {
```
Comment (agentzh) [inline bin/opm]: We should avoid such immediate hash table look ups since when there is a typo in the hash table key here, it is hard to detect and debug.

**Comment by jvanasco** [comment]
Sorry for the delay in response, I was on vacation.

I've pushed the changes, and will address everything above here:

1. why `getcwd`?   IIRC, it is supposed to work better with symlinks.  I started using it in a few spots where paths seems to switch around.

2. whitespace and formatting issues. I'm sorry for this. can you recommend a Perl linter or similar for simple stuff like this?  It's been a very long time since I've used perl.

3. the hash table lookup isn't immediate - it happens in another function several hundred lines away.  in order for me to use the existing logic of `build-install` and `upload` and not duplicate code blocks in-line, I had to decompose some functions into re-usable components (and use a new per-installable context object). 

Looking at this with some fresh eyes (and a little more familiarity with Perl), in the current changeset I also altered a few things.

I didn't quite understand how/why some things originally worked with the chdirs -- but now do (I think) and so I tried to more clearly convey what is going on.  The functions now note the "start" directory before chdiring, and then explicitly return to that 'start' directory (instead of the "install" directory).  This loosely emulates a context manager, and makes chaining or stacking functions a lot easier.

**Comment by agentzh** [comment]
@jvanasco Some comments:

1. I think it's confusing to use both `cwd` and `getcwd` at the same time, especially when they *could* be different. We need to pick just one consistently.
2. I never trust tools to automatically format my code though I'm a big fan of creating tools to report any offenders.
3. You put the hash lookup expression directly in an `if` statement condition which makes typos in the hash keys extremely hard to detect. I think you should always explicitly test the definedness of the values of those hash table keys before testing the boolean value in the `if` statement conditions. This makes potential typos in the hash keys nowhere to hide.

**Comment by jvanasco** [comment]
1. I think getcwd might be better than cwd, because it tries to use the system links and falls back on cwd.

2. I was hoping you knew of a tool like flake8 in python or jshint for javascript.  they simply analyze your code and point out things like whitespace issues, style mistakes from the common formats, possible scope issues, etc.  in my experience, they catch 90% of issues in a PR (from me or to me).

3. Got it on the hash key issue.  I will address!

**Comment by agentzh** [comment]
@jvanasco Yes, Perl has similar tools as well. You can dig them up if you like.

We do prepare our own static analyses toolsets for OpenResty's C components and Lua components. But not yet for Perl.

**Comment by ktalebian** [comment]
@agentzh @jvanasco what's the status on this pr?

**Comment by jvanasco** [comment]
@ktalebian I am closing this PR. There were several changes to OpenResty in the months after I first wrote it, which would have required far too much work to rebase or merge with, and I never had time to revisit this since.  someone else will need to create this functionality.
